engine/table/src/main/java/io/deephaven/engine/table/impl/hierarchical/TreeTableFilter.java

```diff
@@ -0,0 +1,579 @@
+            final boolean usePrev,
+            @NotNull final RowSequence rowsToParent) {
+        if (rowsToParent.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        final Map<Object, RowSetBuilderSequential> parentIdToChildRows =
+                new LinkedHashMap<>(rowsToParent.intSize());
+        try (final ChunkSource.GetContext parentIdGetContext = parentIdSource.makeGetContext(CHUNK_SIZE);
+                final ChunkBoxer.BoxerKernel boxer = ChunkBoxer.getBoxer(parentIdSource.getChunkType(), CHUNK_SIZE);
+                final RowSequence.Iterator childRowsIter = rowsToParent.getRowSequenceIterator()) {
+            final MutableInt chunkOffset = new MutableInt();
+            while (childRowsIter.hasMore()) {
+                final RowSequence chunkChildRows = childRowsIter.getNextRowSequenceWithLength(CHUNK_SIZE);
+                final ObjectChunk<?, ? extends Values> parentIds =
+                        getIds(usePrev, parentIdSource, parentIdGetContext, boxer, chunkChildRows);
+                chunkOffset.setValue(0);
+                chunkChildRows.forAllRowKeys((final long childRowKey) -> {
+                    final Object parentId = parentIds.get(chunkOffset.getAndIncrement());
+                    if (parentId != null) {
+                        parentIdToChildRows.computeIfAbsent(parentId,
+                                pid -> RowSetFactory.builderSequential()).appendKey(childRowKey);
+                    }
+                });
+            }
+        }
+        return parentIdToChildRows;
+    }
+
+    private void shiftParentIdToChildRows(@NotNull final RowSetShiftData upstreamShifts) {
+        try (final ChunkSource.GetContext parentIdGetContext = parentIdSource.makeGetContext(CHUNK_SIZE);
+                final ChunkBoxer.BoxerKernel boxer =
+                        ChunkBoxer.getBoxer(parentIdSource.getChunkType(), CHUNK_SIZE)) {
+            final Set<Object> affectedParents = new HashSet<>();
+            upstreamShifts.apply((final long beginRange, final long endRange, final long shiftDelta) -> {
+                try (final RowSequence affectedRows = resultRows.subSetByKeyRange(beginRange, endRange);
+                        final RowSequence.Iterator affectedRowsIter = affectedRows.getRowSequenceIterator()) {
+                    while (affectedRowsIter.hasMore()) {
+                        final RowSequence chunkAffectedRows =
+                                affectedRowsIter.getNextRowSequenceWithLength(CHUNK_SIZE);
+                        final ObjectChunk<?, ? extends Values> parentIds =
+                                getIds(true, parentIdSource, parentIdGetContext, boxer, chunkAffectedRows);
+                        new ObjectChunkIterator<>(parentIds).forEachRemaining(affectedParents::add);
+                    }
+                    affectedParents.forEach((final Object parentId) -> RowSetShiftData.applyShift(
```

Comment:
I'm not sure I understand your suggestion well enough to implement it. Can you try to say it a different way, or provide some pseudocode?

Reply:
If you look at e.g. io.deephaven.engine.rowset.RowSetShiftData#apply(io.deephaven.engine.rowset.WritableRowSet) we don't actually apply the individual shifts to an Index, we accumulate into a builder then apply at the end.  If you do the shifts in either ascending or descending strict order; you'll blow up because of overwrites.  If you do them in the "proper" order that apply suggests, then we'll end up inserting/removing things from the middle of an index very often.

The solution we need is to write down what we need to add/remove; then do it all at once.